vllm/model_executor/layers/sampler.py

```diff
@@ -60,12 +82,21 @@ def forward(
         assert logits is not None
         _, vocab_size = logits.shape
 
-        logits = _apply_min_tokens_penalty(logits, sampling_metadata)
-
         # Prepare sampling tensors with pinned memory to avoid blocking.
-        (sampling_tensors, do_penalties, do_top_p_top_k,
-         do_min_p) = SamplingTensors.from_sampling_metadata(
-             sampling_metadata, vocab_size, logits.device, logits.dtype)
+        if not sampling_metadata.reuse_sampling_tensors:
+            self._init_sampling_tensors(logits, sampling_metadata)
+        elif self._do_penalties:
+            # In this case, we depend on the output tokens
+            # TODO: Check with Cade if this is needed for spec tokens
```

Comment:
Can we add more to this comment? This case seems to be, when in spec decode and `do_penalties` is True. Why is this case special ?

Reply:
I'm having a hard time understanding the question -- let's chat offline if it's faster.

Reply:
Added a better explanation